File: ACM/Acwing/1752/1752.cpp
#include <algorithm>
#include <cstdio>
using namespace std;
struct node
{
    int l, r, cov, len;
} t[800005];
int n, cnt, tot, q[200005], l[100005], r[100005], ans;
void build(int i, int l, int r)
{
    t[i].l = l;
    t[i].r = r;
    t[i].len = 0;
    if (l == r)
        return;
    int mid = l + r >> 1;
    build(i << 1, l, mid);
    build(i << 1 | 1, mid + 1, r);
} //建树
void update(int i, int l, int r, int v)
{
    if (l == t[i].l && t[i].r == r)
    {
        t[i].cov += v;
        if (t[i].cov == 1)
            t[i].len = q[t[i].r + 1] - q[t[i].l]; //对还未覆盖的区间进行修改
        if (!t[i].cov) //如果该区间为被覆盖，就从下面的子节点上传被覆盖的长度
            if (l == r)
                t[i].len = 0; //前面说到的如果说到了叶节点就不能上传
            else
                t[i].len = t[i << 1].len + t[i << 1 | 1].len;
        return;
    }
    int mid = t[i].l + t[i].r >> 1;
    if (r <= mid)
        update(i << 1, l, r, v);
    else if (l > mid)
        update(i << 1 | 1, l, r, v);
    else
        update(i << 1, l, mid, v), update(i << 1 | 1, mid + 1, r, v);
    if (!t[i].cov)
        t[i].len = t[i << 1].len + t[i << 1 | 1].len; //上传
}
int main()
{
    scanf("%d", &n);
    for (int i = 1; i <= n; ++i)
    {
        scanf("%d %d", &l[i], &r[i]);
        q[++tot] = l[i];
        q[++tot] = r[i]; //预存
    }
    sort(q + 1, q + tot + 1);                     //排序
    int cnt = unique(q + 1, q + tot + 1) - q - 1; //去重
    build(1, 1, cnt - 1);
    for (int i = 1; i <= n; i++)
    {
        l[i] = lower_bound(q + 1, q + cnt + 1, l[i]) - q;
        r[i] = lower_bound(q + 1, q + cnt + 1, r[i]) - q - 1;
        update(1, l[i], r[i], 1);
    } //离散化同时处理出最后的情况
    for (int i = 1; i <= n; i++)
    {
        update(1, l[i], r[i], -1);
        ans = max(ans, t[1].len);
        update(1, l[i], r[i], 1);
    } //枚举所有的情况并求出最大的答案
    printf("%d\n", ans);
    return 0;
}

File: ACM/Acwing/1752/1752.py
'''
Author: NEFU AB-IN
Date: 2022-02-08 08:59:37
FilePath: \ACM\Acwing\1752.py
LastEditTime: 2022-02-09 11:29:37
'''

#扫描线求区间最大覆盖
#学习地址：https://ncc79601.blog.luogu.org/scan-line

ls = lambda p: p << 1
rs = lambda p: p << 1 | 1


class Node(object):
    def __init__(self, l, r) -> None:
        self.l = l
        self.r = r
        self.len = 0  # 区间内被截的长度
        self.tag = 0  # 被完全覆盖的次数


N = int(1010)
tr = [Node(0, 0) for _ in range(N << 3)]
lst = []
xs = []
L = [0 for _ in range(N)]
R = [0 for _ in range(N)]


def find(x):
    l = 0
    r = len(xs) - 1
    while l < r:
        mid = l + r >> 1
        if xs[mid] >= x:
            r = mid
        else:
            l = mid + 1
    return r


def pushup(p):
    if tr[p].tag:  #被覆盖过，这个区间是满的，覆盖长度就是这个区间的长度
        tr[p].len = (xs[tr[p].r + 1] - xs[tr[p].l])  #更新长度
    else:  #否则
        tr[p].len = tr[ls(p)].len + tr[rs(p)].len  #合并儿子信息


def build(p, l, r):
    tr[p] = Node(l, r)
    if l == r:
        return
    mid = l + r >> 1
    build(ls(p), l, mid)
    build(rs(p), mid + 1, r)
    pushup(p)


#l, r 是固定的，二分的永远是tr[p].l和tr[p].r
def update(p, l, r, d):
    if l <= tr[p].l and tr[p].r <= r:
        tr[p].tag += d
        pushup(p)
        return
    #pushdown(p)
    mid = tr[p].l + tr[p].r >> 1
    if l <= mid:
        update(ls(p), l, r, d)
    if mid < r:
        update(rs(p), l, r, d)
    pushup(p)


if __name__ == "__main__":
    n = int(input())
    for i in range(n):
        l, r = map(int, input().split())
        xs.append(l)
        xs.append(r)
        lst.append([l, r])
    xs = list(set(xs))
    xs.sort()
    build(1, 0, len(xs) - 1)
    for i in range(n):
        L[i] = find(lst[i][0])
        R[i] = find(lst[i][1])
        update(1, L[i], R[i] - 1, 1)
    res = 0
    for i in range(n):
        update(1, L[i], R[i] - 1, -1)
        res = max(res, tr[1].len)
        update(1, L[i], R[i] - 1, 1)
    print(res)
